server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -169,54 +170,55 @@ DocumentParser documentParser() {
-            throw e;
-        }
-
-        if (updatedMapper == null) {
-            return false;
-        }
-
-        boolean requireRefresh = false;
-
-        assertMappingVersion(currentIndexMetadata, newIndexMetadata, updatedMapper);
-
-        MappingMetadata mappingMetadata = newIndexMetadata.mapping();
-        CompressedXContent incomingMappingSource = mappingMetadata.source();
-
-        String op = mapper != null ? "updated" : "added";
-        if (logger.isDebugEnabled() && incomingMappingSource.compressed().length < 512) {
-            logger.debug("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
-        } else if (logger.isTraceEnabled()) {
-            logger.trace("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
-        } else {
-            logger.debug("[{}] {} mapping (source suppressed due to length, use TRACE level if needed)",
-                index(), op);
+        MappingMetadata newMappingMetadata = newIndexMetadata.mapping();
+        if (newMappingMetadata != null) {
+            String type = newMappingMetadata.type();
+            CompressedXContent incomingMappingSource = newMappingMetadata.source();
+            Mapping incomingMapping = parseMappings(type, incomingMappingSource);
+            DocumentMapper previousMapper;
+            synchronized (this) {
+                previousMapper = this.mapper;
+                assertRefreshIsNotNeeded(previousMapper, type, incomingMappingSource, incomingMapping);
+                this.mapper = newDocumentMapper(incomingMapping, MergeReason.MAPPING_RECOVERY);
+            }
+            String op = previousMapper != null ? "updated" : "added";
+            if (logger.isDebugEnabled() && incomingMappingSource.compressed().length < 512) {
+                logger.debug("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
+            } else if (logger.isTraceEnabled()) {
+                logger.trace("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
+            } else {
+                logger.debug("[{}] {} mapping (source suppressed due to length, use TRACE level if needed)",
+                    index(), op);
+            }
         }
+    }
 
-        // refresh mapping can happen when the parsing/merging of the mapping from the metadata doesn't result in the same
-        // mapping, in this case, we send to the master to refresh its own version of the mappings (to conform with the
-        // merge version of it, which it does when refreshing the mappings), and warn log it.
-        if (documentMapper().mappingSource().equals(incomingMappingSource) == false) {
-            logger.debug("[{}] parsed mapping, and got different sources\noriginal:\n{}\nparsed:\n{}",
-                index(), incomingMappingSource, documentMapper().mappingSource());
-
-            requireRefresh = true;
+    private void assertRefreshIsNotNeeded(DocumentMapper currentMapper,
```

Comment:
I prefer to make this a method that returns a boolean (return true), and then call that as `assert assertRefreshIsNotNeeded(...)`.  Makes it clearer that this method is not  even called  when assertions are not enabled

Reply:
makes sense to me, I will make this change

Reply:
having tried this one, it makes things slightly more complicated because I need to carry the message out of the method, which depends on the mappings that are being parsed as part of the method itself. I would be leaning towards leaving it as-is

Reply:
I think Yannick means to leave the method as-is (with its own assertions) just make it always return `true`. See e.g. `ReplicationTracker#invariant()`.

Reply:
thanks for explaining that David, I get it now. I made that change.